test/common/buffer/owned_impl_test.cc

```diff
@@ -1263,6 +1265,120 @@ TEST_F(OwnedImplTest, FrontSlice) {
   EXPECT_EQ(1, buffer.frontSlice().len_);
 }
 
+TEST(BufferMemoryAccountTest, ManagesAccountBalance) {
+  std::shared_ptr<BufferMemoryAccount> account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer;
+  buffer.bindAccount(account);
+  ASSERT_EQ(account->balance(), 0);
+
+  // Check the balance increases as expected.
+  {
+    // New slice created
+    buffer.add("Hello");
+    EXPECT_EQ(account->balance(), 4096);
+
+    // Should just be added to existing slice.
+    buffer.add(" World!");
+    EXPECT_EQ(account->balance(), 4096);
+
+    // Trigger new slice creation with add.
+    const std::string long_string(4096, 'a');
+    buffer.add(long_string);
+    EXPECT_EQ(account->balance(), 8192);
+
+    // AppendForTest also adds new slice.
+    buffer.appendSliceForTest("Extra Slice");
+    EXPECT_EQ(account->balance(), 12288);
+  }
+
+  // Check the balance drains as slices are consumed.
+  {
+    // Shouldn't trigger slice free yet
+    buffer.drain(4095);
+    EXPECT_EQ(account->balance(), 12288);
+
+    // Trigger slice reclaim.
+    buffer.drain(1);
+    EXPECT_EQ(account->balance(), 8192);
+
+    // Reclaim next slice
+    buffer.drain(std::string("Hello World!").length());
+    EXPECT_EQ(account->balance(), 4096);
+
+    // Reclaim remaining
+    buffer.drain(std::string("Extra Slice").length());
+    EXPECT_EQ(account->balance(), 0);
+  }
+}
+
+TEST(BufferMemoryAccountTest, BufferAccountsForUnownedSliceMovedInto) {
```

Comment:
Add a similar test that uses addBufferFragment, verify that an account is not associated with the buffer fragments as they are moved.

Reply:
Done.